### ISSUE TITLE
Search dates

### ISSUE THREAD

**Issue by Captainkirkdawson** [FreeUKGen/MyopicVicar#2278]
A review of the distinct search dates shows that the code that takes the event date from an entry does not function correctly with lower case or uppercase months. We have 277,293 distinct search dates yet there are only 219,000 distinct date in 600 years.
We have dates like "29-Mar-02", "1893-apr-02", 1893-July-30", "1893-February-05", "1891-JUN-15", 

Full file uploaded to gdrive
https://drive.google.com/open?id=1K_RhemeZcpe2BydLMK4bgBd_nx01vtdm

**Comment by Captainkirkdawson** [comment]
An example of impact is #1310

**Comment by Captainkirkdawson** [comment]
Note "1893-July-30", "1893-February-05" are valid formats

**Comment by Captainkirkdawson** [comment]
Have closed the other search date stories in favour of this one.
Having dug into the search record code for search date calculation it is clear that only correctly formatted dates will be correctly processed, ie. dates formatted as dd Mmm yyyy. 
The year MAY include a * or _ in the 4th digit and a 5  is generated ie 16 Mar 187* will be treated as 16 Mar 1875. 
The year MAY include __ or * after 2 digits and a 50 is generated ie 16 Mar 18* will be treated as 16 Mar 1850.
The format of the month MUST be Mmm ie first letter capital and 2 lower case letters. Any other format all lower case or all upper case generates a string that will NOT compare correctly to the start and end years in a search.
The day is can be anything eg 999 and this can result in an incorrect comparison if the month is Dec.

Thus the validation done by the processor and on line edit must be tightened to ensure only correctly formatted dates go through to the creation of a search record. In addition the database dates must be revalidated

**Comment by Sherlock21** [comment]
I agree  with your logic completely. ( not meant to be patronising!)

**Comment by Captainkirkdawson** [comment]
Tightened date validation has been applied to test3.
@SteveBiggs  can you please test.
The text for Transcribers on test3 has been adjusted to reflect this change. We may wish to give it wide publicity when released.
@AlOneill please note
**If part or all of the date is missing or illegible**, we make use of parts of our Uncertain Character Format (UCF) as follows:

If the day is missing, just enter a month and year — Jun 1783
If the day is unclear, use  * (asterisk) for an unreadable day —  * Jul 1746
If the month is missing or illegible, enter dd * yyyy — 09 * 1708
If the day and month are missing, just enter the year — 1761
If only the year is legible, put * * yyyy — * * 1761
If the year is missing, then enter the most likely year from the position in the register as appropriate together with a note on the estimation in the Notes field. You may use the * (asterisk) in the 3rd or 4th digit of the year or the _ (underscore) in the 4th digit and the underscore in both the 3rd and 4th digits. Do not use any other UCF (Uncertain Character Format) for the year — 09 Feb 185* , 20 Jun 178_, * * 18*, 15 * 19__

Once deployed we need to consider how best to purge the database of bad search dates.. I am tempted to  do a file by file reprocessing by userid. Comments appreciated..

**Comment by edickens** [comment]
I think in the notes that we need to make clear that there must be a Year.  Just an Asterisk will be rejected, even if it is a blank entry in the register.

In your paragraph in italics, if this is to go in the Help, then include examples of year only with UCF - 18*, 182_

I think the Do not use examples needs to be a new paragraph.

This file OKIEVIBA (Pedlar) in Parish Register of Evie in Evie of OKI has a lot of UCF in dates, but all except one blank entry, where there is just a Reg No, are correct.  A good test.

When you reprocess, warn transcribers first as an email gets sent to them if they are active.  Will you lock the file if there have been changes or just leave the errors with the file as "Files with errors".

**Comment by edickens** [comment]
It is probably much too late now, but would a new field of "Search year" make processing easier and quicker?  Just extract four digit years, with zero or five for UCF numbers.  This is not displayed, but the only field used in the search algorithm.  Even split years are just four digits being the modern year.  This is what the researcher is thinking when they enter years in the search screen.
Even if you don't use this new field now, if you are going to reprocess every file, then now is the time to create this field, even if you decide never to use it.

**Comment by Sherlock21** [comment]
As you say, Eric D, it would have been a god dear if we had started with it, but perhaps it is too late in the day now?

Such a year date would need to be reset every time the transcriber altered the fill date or course,ANS would your create such a "processing year" field for Birth Date and for Death Date as they are (theoretically) might just as likely to be with the same errors?

**Comment by edickens** [comment]
Yes EricB.  I was just thinking of the same problem.  I think one field, but Birth date and Death date would take precedence over Baptism and Burial if both are present.  That is because that is the year that the researcher is thinking.

**Comment by Sherlock21** [comment]
I did a little test of online edit of Burial Date - 
I changed a correct date into one with MONTH in lower case. and it was correctly rejected.
changed it into a Mmm style and that was accepted OK

repeated with Death Date


that failed correctly, then passed correctly when I used the Mmm format.

One little thing: the fail message just simply says:  ["Death date Invalid date”] 

I know what it failed for would it be more use to the transcriber if the reason for failure actually says: 
 ["Death date Invalid format date"]

On 19 May 2020, at 00:16, Kirk Dawson <notifications@github.com> wrote:


Tightened date validation has been applied to test3.
@SteveBiggs <https://github.com/SteveBiggs> can you please test.
The text for Transcribers on test3 has been adjusted to reflect this change. We may wish to give it wide publicity when released.
@AlOneill <https://github.com/AlOneill> please note
If part or all of the date is missing or illegible, we make use of parts of our Uncertain Character Format (UCF) as follows:

If the day is missing, just enter a month and year — Jun 1783
If the day is unclear, use * (asterisk) for an unreadable day — * Jul 1746
If the month is missing or illegible, enter dd * yyyy — 09 * 1708
If the day and month are missing, just enter the year — 1761
If only the year is legible, put * * yyyy — * * 1761
If the year is missing, then enter the most likely year from the position in the register as appropriate together with a note on the estimation in the Notes field. You may use the * (asterisk) in the 3rd or 4th digit of the year or the _ (underscore) in the 4th digit and the underscore in both the 3rd and 4th digits. Do not use any other UCF (Uncertain Character Format) for the year — 09 Feb 185* , 20 Jun 178_, * * 18*, 15 * 19__

Once deployed we need to consider how best to purge the database of bad search dates.. I am tempted to do a file by file reprocessing by userid. Comments appreciated..

**Comment by edickens** [comment]
A thought on the Mmm rule.  I'm sure that a lot of months will not fit this rule, mostly being all uppercase or all lowercase.  Can it not be case insensitive?  It should be three characters only.

**Comment by Captainkirkdawson** [comment]
Text is on refinery see
https://test3.freereg.org.uk/cms/information-for-transcribers/entering-data-from-registers-draft
@edickens Permitting lower case and upper case month would require a change to the search code and not just validation. (The reason we got into part of this mess is that the validation DID do as you suggest but the search code currently expect Mmm)
@edickens @Sherlock21 we likely did make a bad choice on day 1 by making the search date a date formatted string. This means we convert the start and end dates to date formatted string to do a comparison.   It would in hindsight have been better to create an integer search year and do an integer comparison with start and end years. Much more efficient. But we did not. Changing now would be a MAJOR task (3-4 months) especially as we use the search date as part of our sharding key which distributes records across the servers. The database cleanup would not be such a large task as we would only have to update search dates identified by looking at the original entry and DELETING the search record. The TRANSCRIBER/ COORDINATOR would have to do the correction.
PS the effect of this formatting error is NOT universal ie only ill formatted records for the END year of a search range are omitted.

**Comment by AlOneill** [comment]
@Captainkirkdawson Noted. I will endeavour to make the text regarding the year a little more digestible (!). I will also update on Flexible CSV page — a  rare example of (nearly) duplicated information.

**Comment by Sherlock21** [comment]
As you say, Kirk, it would have been better to do things differently - but we didn't.
 
SO now i support you in just doing the easiest fix to the data, so  the search works, as you suggest.

**Comment by SteveBiggs** [comment]
I'm very concerned by your comment @Captainkirkdawson that the TRANSCRIBER/ COORDINATOR would have to do the correction. As you know, most of the transcribers of existing files are no longer contactable and many counties do not have an active coordinator so these corrections will not likely be done and we will potentially lose a lot of searchable records - or am I misunderstanding?

**Comment by Captainkirkdawson** [comment]
@SteveBiggs The only automatic fix would be for lower and upper case 3 character month names and long month names. This would still leave manual corrections for every thing else. The alternative is to leave the database as is. As noted above the error affects only those records that would have satisfied the END year.
Writing testing and running the correction is likely to take 2-3 months elapse time.

**Comment by Sherlock21** [comment]
I know that it won't be me who does it, but I think its basically wrong to knowingly have records in the database, which we can't access because of situations like these.
SO I suggest we divide the fix be two

first, list all records that need correcting, per County/ Place/ Church, and pass them the the current active Cc to fix themselves, or get the owner to do so.

Second, all those still remaining, apply just the Month reformat, as Kist describes above.. Even this would be better than nothing.

Incidentally, does anyone know  how many records are we talking about?

**Comment by AlOneill** [comment]
I'm trying to understand exactly what has changed in the instructions for a missing year: putting a note in the Notes field is clear enough, but entering the estimated year "as appropriate" will be seen as vague as compared to the current instructions. Are we saying do or do not put a ? after the estimate?

*If* the ? is no longer valid, then we need to be explicit about the change as it is part of the current instructions and transcribers will need to know.

**Comment by Captainkirkdawson** [comment]
@AlOneill The ? on the year is the killer at this point. The estimation is not new it was always there I believe

**Comment by AlOneill** [comment]
The information for a missing year will be:

* if the year is missing, then enter the most likely year from the position in the register, in the form yyyy or yyyy/(0)y, as appropriate, together with a note on the estimation in the Notes field, but **do not** append a ? to an estimated year

I will also add a note to the UCF section (on Dealing with uncertainty) for transcribers that refers them to the restrictions specific to dates.

**Comment by Captainkirkdawson** [comment]
@AlOneill OK by me

**Comment by Captainkirkdawson** [comment]
For scrum review. There are a number of ways forward
a) Have the validator impose strict control Mmm and leave existing records as is. Pro; Minimal work. Con: leaves database inconsistent with new validation hence reprocessing or update will cause sudden failure and records with long month name or case difference will not be retrieved if year matches end year of search
b)Have the validator impose strict control Mmm and reprocess all records with incorrectly formatted records using the current search date creator. Pro: Records will be retrieved. Con: Considerable (months) programming, processing management. Considerable (months) work for coordinators and transcribers to correct new errors. Eliminates use of ?
c)Have the search date creator and validator both accept long and short case insensitive months and reprocess all records with incorrectly formatted records using the new search date creator. Pro; Most records will be retrieved. Con: Considerable (months) programming, processing management. Limited work for Coordinators  who need to deal with odd cases only.
d)Change the search criteria for the end date from lte year-12-31 to lt (year+1)-01-01. Pro; Both correctly and incorrectly formatted records for the year will be retrieved and little work required of technical and none for the coordinator. No requirement to change documentation. ? can still be used to represent uncertainty (In day, month or year). Con: Leaves old records in the database poorly formatted and continues to create poorly formatted records but at least its consistent.
e)Have the search date creator and validator both accept long and short case insensitive months. Change the search criteria for the end date from lte year-12-31 to lt (year+1)-01-01. Pro; New and existing   correctly and incorrectly formatted records will be retrieved and little work required of technical or coordinator. New records are correctly formatted Con: Leaves old records in the database poorly formatted ie database inconsistent
There are other combinations of the preceding options
The simplest by far are d) or e)

**Comment by Captainkirkdawson** [comment]
@AlOneill You may want to hold fire until after a decision has been made as it will change what has been written and edited

**Comment by AlOneill** [comment]
@Captainkirkdawson Noted. The draft will remain a draft until decision is made.

**Comment by suffolkroots** [comment]
@Captainkirkdawson said---The year MAY include a * or _ in the 4th digit and a 5 is generated ie 16 Mar 187* will be treated as 16 Mar 1875.
The year MAY include __ or * after 2 digits and a 50 is generated ie 16 Mar 18* will be treated as 16 Mar 1850.

Are such dates to be presented to researcher as eg 16 Mar 1875 16 Mar 1850? if so an explanatory note wll be needed re <5> <50> years.

**Comment by Captainkirkdawson** [comment]
The situation noted by @suffolkroots is NOT new the system has done that since day 1. The date presented to the researcher remains unaltered. 
This story does NOT propose to change the performance of the system. It simply documents what happens when those situations occur.

**Comment by SteveBiggs** [comment]
As @Captainkirkdawson says the date presented to the researcher will be as transcribed, e.g. 16 Mar 18*. However, note that it will only be returned by a date range search that includes 1850, so a range of 1800 to 1850 will return it while a range of 1800 to 1849 will not.

**Comment by Captainkirkdawson** [comment]
Still awaiting decision on way ahead see post 6 comments above.

**Comment by Sherlock21** [comment]
Re your last post:

do we have any idea how many records would need reprocessing?
I would have thought that's the start point to then pick an answer to your question/suggestion on how to tackle the correcting job.

But I would have thought:
List record userID; file name; field name needing correction: and listed: By County / Transcriber ID.

**Comment by richpomfret** [comment]
Agreed on option 'e' in the scrum.

**Comment by PatReynolds** [comment]
Asking as a family historian: would it be good practice to always make the end date the year after the suspected end date (e.g. I have a September quarter 1895 death from FreeBMD.  Would it be wise to put in the range to search as 1895 - 1896 rather than as 1895-1895? Or would that only work if looking for John Smith and not for a name that is unlikely to occur every year?

**Comment by Captainkirkdawson** [comment]
Do not understand 1895 -1895 is 1 Jan 1895 to 1 Jan 1896; 1895-1896 becomes 1 Jan 1895 - 1 Jan 1897

**Comment by SteveBiggs** [comment]
I think most researchers would add a year or two both before the start and after the end year expected to ensure they catch the record looked for - I certainly do.
